File: src/routes/pengumuman.ts
import { Hono } from "hono";
import { Env, Pengumuman } from "../types";
import { successResponse, errorResponse } from "../utils/response";
import { authMiddleware } from "../middleware/auth";

const pengumumanRouter = new Hono<{ Bindings: Env }>();

/**
 * Helper to generate URL-friendly slug
 */
function slugify(text: string): string {
    return text
        .toString()
        .toLowerCase()
        .trim()
        .replace(/\s+/g, "-")
        .replace(/[^\w\-]+/g, "")
        .replace(/\-\-+/g, "-");
}

/**
 * GET /api/pengumuman (Public)
 * Support query: ?kategori=kegiatan|oprec|ultah&search=...&limit=10&page=1
 */
pengumumanRouter.get("/", async (c) => {
    const kategori = c.req.query("kategori");
    const search = c.req.query("search");
    const limit = parseInt(c.req.query("limit") || "10", 10);
    const page = parseInt(c.req.query("page") || "1", 10);
    const offset = (page - 1) * limit;

    let query = "SELECT * FROM pengumuman WHERE 1=1";
    const params: any[] = [];

    if (kategori && kategori !== "all") {
        query += " AND category = ?";
        params.push(kategori);
    }

    if (search) {
        query += " AND (title LIKE ? OR content LIKE ?)";
        params.push(`%${search}%`, `%${search}%`);
    }

    query += " ORDER BY date_published DESC, created_at DESC LIMIT ? OFFSET ?";
    params.push(limit, offset);

    const { results } = await c.env.DB.prepare(query).bind(...params).all<Pengumuman>();

    // Count total for pagination
    let countQuery = "SELECT COUNT(*) as total FROM pengumuman WHERE 1=1";
    const countParams: any[] = [];
    if (kategori && kategori !== "all") {
        countQuery += " AND category = ?";
        countParams.push(kategori);
    }
    if (search) {
        countQuery += " AND (title LIKE ? OR content LIKE ?)";
        countParams.push(`%${search}%`, `%${search}%`);
    }
    const countResult = await c.env.DB.prepare(countQuery).bind(...countParams).first<{ total: number }>();
    const total = countResult?.total || 0;

    return c.json(successResponse({
        items: results || [],
        pagination: {
            page,
            limit,
            total,
            totalPages: Math.ceil(total / limit)
        }
    }));
});

/**
 * GET /api/pengumuman/:idOrSlug (Public)
 */
pengumumanRouter.get("/:idOrSlug", async (c) => {
    const idOrSlug = c.req.param("idOrSlug");

    const item = await c.env.DB.prepare(
        "SELECT * FROM pengumuman WHERE id = ? OR slug = ?"
    ).bind(idOrSlug, idOrSlug).first<Pengumuman>();

    if (!item) {
        return c.json(errorResponse("Pengumuman tidak ditemukan"), 404);
    }

    // Increment views asynchronously
    c.executionCtx.waitUntil(
        c.env.DB.prepare("UPDATE pengumuman SET views = views + 1 WHERE id = ?").bind(item.id).run()
    );

    return c.json(successResponse(item));
});

/**
 * POST /api/pengumuman (Protected - Admin)
 */
pengumumanRouter.post("/", authMiddleware, async (c) => {
    const body = await c.req.json<{
        title?: string;
        category?: 'kegiatan' | 'oprec' | 'ultah' | 'lainnya';
        content?: string;
        image_url?: string;
        date_published?: string;
        author?: string;
    }>();

    if (!body.title || !body.category || !body.content || !body.date_published) {
        return c.json(errorResponse("Field title, category, content, dan date_published wajib diisi"), 400);
    }

    const id = `ann_${Date.now()}_${Math.random().toString(36).substring(2, 7)}`;
    const baseSlug = slugify(body.title);
    const slug = `${baseSlug}-${Math.random().toString(36).substring(2, 6)}`;
    const author = body.author || "Pengurus PMK";
    const imageUrl = body.image_url || null;

    await c.env.DB.prepare(
        `INSERT INTO pengumuman (id, title, slug, category, content, image_url, date_published, author)
         VALUES (?, ?, ?, ?, ?, ?, ?, ?)`
    ).bind(id, body.title, slug, body.category, body.content, imageUrl, body.date_published, author).run();

    const created = await c.env.DB.prepare("SELECT * FROM pengumuman WHERE id = ?").bind(id).first<Pengumuman>();
    return c.json(successResponse(created, "Pengumuman berhasil ditambahkan"), 201);
});

/**
 * PUT /api/pengumuman/:id (Protected - Admin)
 */
pengumumanRouter.put("/:id", authMiddleware, async (c) => {
    const id = c.req.param("id");
    const body = await c.req.json<{
        title?: string;
        category?: 'kegiatan' | 'oprec' | 'ultah' | 'lainnya';
        content?: string;
        image_url?: string;
        date_published?: string;
        author?: string;
    }>();

    const existing = await c.env.DB.prepare("SELECT * FROM pengumuman WHERE id = ?").bind(id).first<Pengumuman>();
    if (!existing) {
        return c.json(errorResponse("Pengumuman tidak ditemukan"), 404);
    }

    const title = body.title ?? existing.title;
    const category = body.category ?? existing.category;
    const content = body.content ?? existing.content;
    const imageUrl = body.image_url !== undefined ? body.image_url : existing.image_url;
    const datePublished = body.date_published ?? existing.date_published;
    const author = body.author ?? existing.author;

    await c.env.DB.prepare(
        `UPDATE pengumuman 
         SET title = ?, category = ?, content = ?, image_url = ?, date_published = ?, author = ?, updated_at = CURRENT_TIMESTAMP
         WHERE id = ?`
    ).bind(title, category, content, imageUrl, datePublished, author, id).run();

    const updated = await c.env.DB.prepare("SELECT * FROM pengumuman WHERE id = ?").bind(id).first<Pengumuman>();
    return c.json(successResponse(updated, "Pengumuman berhasil diperbarui"));
});

/**
 * DELETE /api/pengumuman/:id (Protected - Admin)
 */
pengumumanRouter.delete("/:id", authMiddleware, async (c) => {
    const id = c.req.param("id");

    const existing = await c.env.DB.prepare("SELECT * FROM pengumuman WHERE id = ?").bind(id).first<Pengumuman>();
    if (!existing) {
        return c.json(errorResponse("Pengumuman tidak ditemukan"), 404);
    }

    await c.env.DB.prepare("DELETE FROM pengumuman WHERE id = ?").bind(id).run();
    return c.json(successResponse(null, "Pengumuman berhasil dihapus"));
});

export default pengumumanRouter;

File: routers/pengumuman.py
import re
import time
import random
import string
from typing import Optional
from fastapi import APIRouter, Depends, HTTPException, Query, BackgroundTasks
from sqlalchemy.orm import Session
from sqlalchemy import or_, desc

from database import get_db
import models
import schemas
from utils.auth import get_current_admin

router = APIRouter(prefix="/api/pengumuman", tags=["Pengumuman"])

def slugify(text: str) -> str:
    """Generate slug from title."""
    text = text.lower().strip()
    text = re.sub(r'[^\w\s-]', '', text)
    text = re.sub(r'[\s_-]+', '-', text)
    text = re.sub(r'^-+|-+$', '', text)
    return text

def increment_views_task(pengumuman_id: str, db: Session):
    try:
        item = db.query(models.Pengumuman).filter(models.Pengumuman.id == pengumuman_id).first()
        if item:
            item.views += 1
            db.commit()
    except Exception:
        pass

@router.get("", response_model=schemas.ApiResponse[schemas.PengumumanListResponse])
def get_pengumuman_list(
    kategori: Optional[str] = Query(None, description="Filter: kegiatan | oprec | ultah | lainnya"),
    search: Optional[str] = Query(None, description="Search keyword in title or content"),
    page: int = Query(1, ge=1),
    limit: int = Query(10, ge=1, le=50),
    db: Session = Depends(get_db)
):
    """Get announcements list with filtering, searching, and pagination."""
    query = db.query(models.Pengumuman)

    if kategori and kategori != "all":
        query = query.filter(models.Pengumuman.category == kategori)

    if search:
        search_fmt = f"%{search}%"
        query = query.filter(
            or_(
                models.Pengumuman.title.ilike(search_fmt),
                models.Pengumuman.content.ilike(search_fmt)
            )
        )

    total = query.count()
    items = query.order_by(desc(models.Pengumuman.date_published), desc(models.Pengumuman.created_at))\
                 .offset((page - 1) * limit)\
                 .limit(limit)\
                 .all()

    totalPages = (total + limit - 1) // limit if limit > 0 else 0

    return schemas.ApiResponse(
        success=True,
        data=schemas.PengumumanListResponse(
            items=[schemas.PengumumanOut.model_validate(item) for item in items],
            pagination=schemas.PaginationMeta(
                page=page,
                limit=limit,
                total=total,
                totalPages=totalPages
            )
        )
    )

@router.get("/{id_or_slug}", response_model=schemas.ApiResponse[schemas.PengumumanOut])
def get_pengumuman_detail(
    id_or_slug: str,
    background_tasks: BackgroundTasks,
    db: Session = Depends(get_db)
):
    """Get single announcement detail by ID or Slug."""
    item = db.query(models.Pengumuman).filter(
        or_(models.Pengumuman.id == id_or_slug, models.Pengumuman.slug == id_or_slug)
    ).first()

    if not item:
        raise HTTPException(status_code=404, detail="Pengumuman tidak ditemukan")

    # Increment views in background task
    background_tasks.add_task(increment_views_task, item.id, db)

    return schemas.ApiResponse(
        success=True,
        data=schemas.PengumumanOut.model_validate(item)
    )

@router.post("", response_model=schemas.ApiResponse[schemas.PengumumanOut], status_code=201)
def create_pengumuman(
    payload: schemas.PengumumanCreate,
    current_user: models.User = Depends(get_current_admin),
    db: Session = Depends(get_db)
):
    """Create new announcement (Admin only)."""
    random_suffix = ''.join(random.choices(string.ascii_lowercase + string.digits, k=4))
    item_id = f"ann_{int(time.time())}_{random_suffix}"
    slug = f"{slugify(payload.title)}-{random_suffix}"

    new_item = models.Pengumuman(
        id=item_id,
        title=payload.title,
        slug=slug,
        category=payload.category,
        content=payload.content,
        image_url=payload.image_url,
        date_published=payload.date_published,
        author=payload.author or "Pengurus PMK"
    )

    db.add(new_item)
    db.commit()
    db.refresh(new_item)

    return schemas.ApiResponse(
        success=True,
        data=schemas.PengumumanOut.model_validate(new_item),
        message="Pengumuman berhasil dibuat"
    )

@router.put("/{item_id}", response_model=schemas.ApiResponse[schemas.PengumumanOut])
def update_pengumuman(
    item_id: str,
    payload: schemas.PengumumanUpdate,
    current_user: models.User = Depends(get_current_admin),
    db: Session = Depends(get_db)
):
    """Update announcement (Admin only)."""
    item = db.query(models.Pengumuman).filter(models.Pengumuman.id == item_id).first()
    if not item:
        raise HTTPException(status_code=404, detail="Pengumuman tidak ditemukan")

    if payload.title is not None:
        item.title = payload.title
    if payload.category is not None:
        item.category = payload.category
    if payload.content is not None:
        item.content = payload.content
    if payload.image_url is not None:
        item.image_url = payload.image_url
    if payload.date_published is not None:
        item.date_published = payload.date_published
    if payload.author is not None:
        item.author = payload.author

    db.commit()
    db.refresh(item)

    return schemas.ApiResponse(
        success=True,
        data=schemas.PengumumanOut.model_validate(item),
        message="Pengumuman berhasil diperbarui"
    )

@router.delete("/{item_id}", response_model=schemas.ApiResponse[None])
def delete_pengumuman(
    item_id: str,
    current_user: models.User = Depends(get_current_admin),
    db: Session = Depends(get_db)
):
    """Delete announcement (Admin only)."""
    item = db.query(models.Pengumuman).filter(models.Pengumuman.id == item_id).first()
    if not item:
        raise HTTPException(status_code=404, detail="Pengumuman tidak ditemukan")

    db.delete(item)
    db.commit()

    return schemas.ApiResponse(
        success=True,
        message="Pengumuman berhasil dihapus"
    )
